node/impl/full/multisig.go

```diff
@@ -0,0 +1,271 @@
+package full
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/actors"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/filecoin-project/specs-actors/actors/builtin"
+	init_ "github.com/filecoin-project/specs-actors/actors/builtin/init"
+	samsig "github.com/filecoin-project/specs-actors/actors/builtin/multisig"
+
+	"github.com/ipfs/go-cid"
+	"github.com/minio/blake2b-simd"
+	cbg "github.com/whyrusleeping/cbor-gen"
+
+	"go.uber.org/fx"
+	"golang.org/x/xerrors"
+)
+
+type MsigAPI struct {
+	fx.In
+
+	WalletAPI WalletAPI
+	StateAPI  StateAPI
+	MpoolAPI  MpoolAPI
+}
+
+func (a *MsigAPI) MsigCreate(ctx context.Context, req int64, addrs []address.Address, val types.BigInt, src address.Address, gp types.BigInt) (api.AddressPair, error) {
+
+	lenAddrs := int64(len(addrs))
+
+	if lenAddrs < req {
+		return api.UndefAddressPair, xerrors.Errorf("cannot require signing of more addresses than provided for multisig")
+	}
+
+	if req == 0 {
+		req = lenAddrs
+	}
+
+	if src == address.Undef {
+		defaddr, err := a.WalletAPI.WalletDefaultAddress(ctx)
+		if err != nil {
+			return api.UndefAddressPair, err
+		}
+		src = defaddr
+	}
+
+	if gp == types.EmptyInt {
+		gp = types.NewInt(1)
+	}
+
+	// Set up constructor parameters for multisig
+	msigParams := &samsig.ConstructorParams{
+		Signers:               addrs,
+		NumApprovalsThreshold: req,
+	}
+
+	enc, actErr := actors.SerializeParams(msigParams)
+	if actErr != nil {
+		return api.UndefAddressPair, actErr
+	}
+
+	// new actors are created by invoking 'exec' on the init actor with the constructor params
+	execParams := &init_.ExecParams{
+		CodeCID:           builtin.MultisigActorCodeID,
+		ConstructorParams: enc,
+	}
+
+	enc, actErr = actors.SerializeParams(execParams)
+	if actErr != nil {
+		return api.UndefAddressPair, actErr
+	}
+
+	// now we create the message to send this with
+	msg := types.Message{
+		To:       builtin.InitActorAddr,
+		From:     src,
+		Method:   builtin.MethodsInit.Exec,
+		Params:   enc,
+		GasPrice: gp,
+		GasLimit: 1000000,
+		Value:    val,
+	}
+
+	// send the message out to the network
+	smsg, err := a.MpoolAPI.MpoolPushMessage(ctx, &msg)
+	if err != nil {
+		return api.UndefAddressPair, err
+	}
+
+	// wait for it to get mined into a block
+	wait, err := a.StateAPI.StateWaitMsg(ctx, smsg.Cid())
```

Comment:
In case there's a holdup, adding a statement ("Waiting for the message to be mined") wouldn't hurt.

Reply:
That's a good call. Especially since I included one in `MsigProposal`. 
Added `fmt.Println("send proposal in message: ", smsg.Cid())` to make this more symmetrical to the other one.

Reply:
Added, resolving